File: src/main/java/org/usfirst/irs1318/gamesim/engine/activities/Move.java
package org.usfirst.irs1318.gamesim.engine.activities;

import org.usfirst.irs1318.gamesim.engine.actors.*;
import org.usfirst.irs1318.gamesim.engine.objective.TaskDetails;

public final class Move extends Activity {
    private final Field field;
    private final Direction direction;

    public Move(Actor actor, TaskDetails taskDetails, Field field, Direction direction) {
        super(actor, taskDetails);
        this.field = field;
        this.direction = direction;
    }

    @Override
    public void performSuccessActivity() {
        Location start = actor.getLocation();
        Location destination = field.getDestination(start, direction);
        if (!destination.equals(start)) {
            actor.setLocation(destination);
        }
    }

    @Override
    public void performFailureActivity() {
        // TODO
    }

    public Field getField() {
        return field;
    }

    public Direction getDirection() {
        return direction;
    }
}

File: src/main/scala/org/usfirst/irs1318/gamesim/engine/action/Move.scala
package org.usfirst.irs1318.gamesim.engine.action

import org.usfirst.irs1318.gamesim.engine.GameEngine
import org.usfirst.irs1318.gamesim.engine.actors._

case class Move(name: String, direction: Direction) extends Action {
  override def run(state: GameEngine.State): GameEngine.State = state
    .mapActor(name, _.mapLocation(_.translate(direction)))
}
